File: System-Programming-Fundamentals-5th-Term/Lab4-head/head.pl
#!/usr/bin/perl -T

# Run
# cpan Getopt::Long
# to fetch script dependencies

use strict;
use warnings qw(FATAL all);
use Getopt::Long;

use constant USAGE => "Usage: $0 [-n num-lines] [-c num-bytes] [file...]\n";

my $num_lines;
my $num_chars;

GetOptions("n=i" => \$num_lines, "c=i" => \$num_chars) or die(USAGE);
die("$0: invalid number of lines: '$num_lines'\n") if ($num_lines and $num_lines < 0);
die("$0: invalid number of characters: '$num_chars'\n") if ($num_chars and $num_chars < 0);
exit if $num_lines and $num_lines == 0 or $num_chars and $num_chars == 0;
$num_lines = 10 if not $num_lines and not $num_chars;

my $printed = 0;
my $first_file = 1;

while (@ARGV or $first_file or $printed > 0) {
  if ($printed == 0) {
    if (@ARGV) {
      my $leading_newline = $first_file ? "" : "\n";
      my $filename = $ARGV[0] eq "-" ? "standard input" : $ARGV[0];
      print "$leading_newline==> $filename <==\n";
      undef $first_file;
    }
    else {
      # No arguments = don't print the header, read from stdin
      undef $first_file;
    }
  }

  if ($num_lines and $printed++ < $num_lines) {
    eval {
      my $line = <>;
      print $line;
    }
    or do {
      print $@; # error
      $printed = 0;
    }
  }
  elsif ($num_chars and $printed < $num_chars) {
    eval {
      my $line = <>;
      my $remaining_chars = $num_chars - $printed;
      my $should_print = $remaining_chars < length $line ? $remaining_chars : length $line;
      print substr $line, 0, $should_print;
      $printed += $should_print;
    }
    or do {
      print $@; # error
      $printed = 0;
    }
  }

  if ($num_lines and $printed == $num_lines or $num_chars and $printed == $num_chars or eof) {
    $printed = 0;
    close ARGV;
  }
}

File: System-Programming-Fundamentals-5th-Term/Lab4-head/head.c
#define _POSIX_C_SOURCE 2
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdarg.h>
#include <errno.h>

#define BUF_SIZE 4096

void print_lines_buffered(int fd, unsigned int num_lines) {
  unsigned int printed = 0;

  char inbuf[BUF_SIZE];
  int bytes_read;
  while (printed < num_lines && (bytes_read = read(fd, inbuf, BUF_SIZE)) > 0) {
    unsigned int pos;
    for (pos = 0; pos < bytes_read; ++pos)
      if (inbuf[pos] == '\n' && ++printed == num_lines) break;

    write(STDOUT_FILENO, inbuf, pos);
  }

  write(STDOUT_FILENO, "\n", sizeof("\n"));
}

bool try_parse_uint(const char* str, unsigned int* val) {
  errno = 0;
  char* endptr;
  *val = strtoul(str, &endptr, 10);
  return *str != '-' && errno == 0 && endptr != str && *endptr == '\0';
}

// An fprintf()-like wrapper over write().
// Rationale: the assignment requires we only use raw syscalls to perform IO.
void fdprintf(int fd, const char* fmt, ...) {
  #define MSG_BUF_SIZE 512
  char buf[MSG_BUF_SIZE];
  va_list args;
  va_start(args, fmt);
  int msg_len = vsnprintf(buf, MSG_BUF_SIZE, fmt, args);
  va_end(args);
  write(fd, buf, msg_len < MSG_BUF_SIZE ? msg_len : MSG_BUF_SIZE);
}

void handle_filename_args(int argc, char** argv, int num_lines) {
  // According to POSIX,
  // If multiple file operands are specified, head shall precede the output for each with the header:
  // "\n==> %s <==\n", <pathname>
  // except that the first header written shall not include the initial <newline>.
  #define HEADER_STDIN "\n==> standard input <==\n"
  #define HEADER_FILE_FMT "\n==> %s <==\n"

  bool first_header = true;
  do {
    int header_offset = first_header ? 1 : 0;
    first_header = false;

    if (*argv[optind] == '-') {
      write(STDOUT_FILENO, HEADER_STDIN + header_offset, sizeof(HEADER_STDIN) - header_offset);
      print_lines_buffered(STDIN_FILENO, num_lines);
    }
    else {
      errno = 0;
      int fd = open(argv[optind], O_RDONLY);
      if (fd == -1) {
        const char* error_fmt = "%s: Cannot open %s for reading\n";
        switch (errno) {
          case EACCES:
            error_fmt = "%s: Cannot open %s for reading (access denied)\n";
            break;
          case ENOENT:
            error_fmt = "%s: Cannot open %s for reading (file not found)\n";
            break;
        }
        fdprintf(STDERR_FILENO, error_fmt, argv[0], argv[optind]);
      }
      else {
        fdprintf(STDOUT_FILENO, HEADER_FILE_FMT + header_offset, argv[optind]);
        print_lines_buffered(fd, num_lines);
        close(fd);
      }
    }
  }
  while (++optind < argc);
}

int main(int argc, char** argv) {
  unsigned int num_lines = 10;

  int c;
  while ((c = getopt(argc, argv, "n:")) != EOF) {
    switch (c) {
      case 'n':
        if (!try_parse_uint(optarg, &num_lines)) {
          fdprintf(STDERR_FILENO, "%s: invalid number of lines: '%s'\n", argv[0], optarg);
          return 1;
        }
        break;
      case '?':
        fdprintf(STDERR_FILENO, "Usage: %s [-n num-lines] [file...]\n", argv[0]);
        return 1;
    }
  }

  if (optind == argc)
    print_lines_buffered(STDIN_FILENO, num_lines);
  else
    handle_filename_args(argc, argv, num_lines);

  return 0;
}
